text/0000-experimental_keywords.md

```diff
@@ -0,0 +1,74 @@
+
+It is probable that usage of an experimental keyword after the final syntax has been stabilized will fire a warning that you should move to the final syntax, but this can be decided on a case by case basis for each language feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+From time to time you may see things like `r#$name` in Rust code. These are "experimental keywords".
+
+An experimental keyword represents accessing the "raw operations" of the compiler and language. They have an intentionally funny syntax because experimental keywords are not intended to be used directly in the long term. Instead, they're a way to add features to the language before we've decided on the final syntax for that feature.
+
+Most experimental keywords are given a final syntax intended for general, long-term use, some time after the feature itself becomes available in the Stable compiler. This way the Rust community has an opportunity to try it out and see how it feels, how they use it in practice, and so on. Some experimental keywords aren't ever given a direct syntax of their own, instead they're simply there so that proc-macros can expand to code that uses the keyword.
+
+[And here we'd ideally we'd add at least one or two real examples to the guide once we have some examples of this in practice]
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC alters the language so that `r # $ token` is an accepted token sequence.
+* It names an experimental keyword named `token`.
+* The experimental keyword named _must_ be known to the compiler. If you attempt to name an experimental keyword that the compiler doesn't recognize it is an error.
+* The exact meaning and usage for each experimental keyword depends on the language feature it's associated with, and is not specified here.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* It can slightly hurt Rust's stability story to deliberately introduce new keywords that are intended to become deprecated within a relatively short period of time.
+  * Mitigation 1: Experimental keywords that do become Stable would continue to work even once the final syntax is decided upon and also stabilized. Stable code would not break.
+  * Mitigation 2: The rather unusual `r#$` prefix stands out as "clearly unusual"
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* Alternative: We could select _some other_ sequence of invalid tokens to use as a prefix for experimental keywords.
+  * `r#$` is used because it is close enough to the "raw literal" and "raw string" syntaxes that people are very likely to understand what's going on even if they're not familiar with a specific experimental keyword when they first see it.
```

Comment:
I think this needs a justification for `$` in particular.  Why is `r#$` better than `r#@`, say?  (Not that I think that one's necessarily superior, just that it's not clearly worse either.)

Historically macros have zealously guarded the use of `$` for themselves, and it's not impossible to imagine someone trying to do `r#$mystringliteral` in a macro expansion...

Reply:
Speaking as the RFC author: the current text simply reflects the "first thing that came to mind" during the discussion that lead to this proposal, with not much bike-shedding during that discussion.

Speaking as a user: either way is fine and I have no opinion.

So I'll happily put in whatever people want to decide on for this.

Reply:
`r#$mystring` doesn't work today, and I don't think it will ever work in the future. You can't just stick a `r#` next to a lexed token to turn it into a raw string. Macro's use of `$` shouldn't really matter here.

Reply:
@kennytm Sure, but macro authors use `$` so commonly that it can cause confusion (they might *think* that `r#$foo` can work). So why not reduce that confusion by choosing a different character?

Reply:
well for sure there are plenty of alternatives we could choose.

spellings which currently are tokenization error:

* `r##keyword`
* `r#[keyword]`
* `r#@keyword` / `r#!keyword` / etc
* `r#805209`
* `br#haha_lang_design_go`

Reply:
FWIW, `br#keyword` was also mentioned in #2151.

Reply:
The only question I have while reading this thread is "why?".
Why come up with these contrived monstrosities messing with lexer if pretty much any combination of two existing keywords would work (`do catch` style)?

If you want a generic tool for this (which is, again, excessive in my opinion), make `__experimental` a keyword and use `__experimental other_keyword` as the combination.